2021/day20/day20.java

```diff
@@ -0,0 +1,108 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.io.File;
+
+public class day20 {
+    public static String readAlgorithm(String filename) {
+        String algorithm = null;
+        try {
+            Scanner sc = new Scanner(new File(filename));
+            algorithm = sc.nextLine();
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        }
+        return algorithm;
+    }
+
+    public static ArrayList<ArrayList<Character>> readImage(String filename) {
+        ArrayList<ArrayList<Character>> image = new ArrayList<>();
+        try {
+            Scanner sc = new Scanner(new File(filename));
+            sc.nextLine();
+            sc.nextLine();
+            sc.useDelimiter("");
+            while (sc.hasNextLine()) {
+                image.add(stringToList(sc.nextLine()));
+            }
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        }
+        return image;
+    }
+
+    public static ArrayList<ArrayList<Character>> surroundWithDots(ArrayList<ArrayList<Character>> image) {
```

Comment:
This smells like a (theoretically) great place to use linked lists - you keep on adding things at the beginning and at the end. However, do note that linked lists may not be faster than dynamic arrays in practice, even if they are in theory - read up on cache locality if you're interested in this.

On the other hand, you do need the indices when you do the checking of each pixel, so it's difficult to tell which approach is faster. How about some small benchmarking again?